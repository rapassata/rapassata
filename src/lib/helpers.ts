import {
    AnyShapeTomato,
    ArrayShapedTomato,
    AtomShapedTomato,
    ObjectShapedTomato,
    RecordShapedTomato,
    Tomato,
    TomatoShape,
} from './tomatos';

// T => T | ?T
export type ProcessRequired<T, R> = R extends true ? T : T | undefined;

// Tomato<I, R, S> => I OR ?I based on R
type UnwrapTomato<T> = T extends Tomato<infer I, infer R, infer S> ? ProcessRequired<I, R> : T;

// Tomato<I, R, S> => S
type UnwrapTomatoShape<T> = T extends Tomato<infer T, infer R, infer S> ? S : T;

// Tomato<I, R, S> => Value (recursive)
export type TomatoToValues<T, U = UnwrapTomato<T>, S = UnwrapTomatoShape<T>> = S extends TomatoShape.Atom
    ? U
    : S extends TomatoShape.Array
    ? Array<UnwrapTomato<UnbakeTomato<U>>>
    : { [key in keyof U]: TomatoToValues<UnbakeTomato<U[key]>> };

// (AtomShapedTomato<...X> | ArrayShapedTomato<...X> | ...) => Tomato<...X, S>
export type UnbakeTomato<T> = T extends AtomShapedTomato<infer U, infer R>
    ? Tomato<U, R, TomatoShape.Atom>
    : T extends ArrayShapedTomato<infer U, infer R>
    ? Tomato<U, R, TomatoShape.Array>
    : T extends ObjectShapedTomato<infer U, infer R>
    ? Tomato<U, R, TomatoShape.Object>
    : T extends RecordShapedTomato<infer U, infer R>
    ? Tomato<U, R, TomatoShape.Record>
    : T;

// (AtomShapedTomato<...X> | ArrayShapedTomato<...X> | ...) => Value (recursive)
export type Values<T> = TomatoToValues<UnbakeTomato<T>>;

export type Values2<T, U, S> = TomatoToValues<UnbakeTomato<T>, U, S>;
