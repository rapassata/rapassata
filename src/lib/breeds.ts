import {
    AnyShapeTomato,
    ArrayTomato,
    AtomShapedTomato,
    ObjectTomato,
    TomatoShape,
    Breeds,
    RecordTomato,
} from './tomatos';
import { FlowType } from './validate';
import { isObject } from './helpers';

const baseNode = <B extends { shape: TomatoShape.Atom }, T>(base: B) => {
    const node: AtomShapedTomato<T> = {
        ...base,
        flow: [],
        required: false,
        default: undefined,
        validate: (validate, message = '') => ({
            ...node,
            flow: [...node.flow, { message, validate, type: FlowType.Validate }],
        }),
        require: () => ({ ...node, required: true }),
        defaultTo: x => ({ ...node, default: x }),
    };
    return node;
};

const atomNode = <T>() => {
    const node: AtomShapedTomato<T> = baseNode({
        shape: TomatoShape.Atom,
    });
    return node;
};

export const string: AtomShapedTomato<string> = (() => {
    const node = atomNode<string>();
    return node.validate(x => typeof x === 'string', 'Not a string');
})();

export const any: AtomShapedTomato<any> = (() => {
    return atomNode<any>();
})();

export const number: AtomShapedTomato<number> = (() => {
    const node = atomNode<number>();
    return node.validate(x => typeof x === 'number', 'Not a number');
})();

export const boolean: AtomShapedTomato<boolean> = (() => {
    const node = atomNode<boolean>();
    return node.validate(x => typeof x === 'boolean', 'Not a boolean');
})();

export const array = <T extends AnyShapeTomato>(item: T): ArrayTomato<T> => {
    const node: ArrayTomato<T> = {
        item,
        shape: TomatoShape.Array,
        flow: [],
        required: false,
        default: undefined,
        validate: (validate, message = '') => ({
            ...node,
            flow: [...node.flow, { message, validate, type: FlowType.Validate }],
        }),
        require: () => ({ ...node, required: true }),
        defaultTo: x => ({ ...node, default: x }),
    };
    return node.validate(x => Array.isArray(x), 'Not an array');
};


export const object = <T>(structure: T): ObjectTomato<T> => {
    const node: ObjectTomato<T> = {
        structure,
        shape: TomatoShape.Object,
        flow: [],
        required: false,
        default: undefined,
        validate: (validate, message = '') => ({
            ...node,
            flow: [...node.flow, { message, validate, type: FlowType.Validate }],
        }),
        require: () => ({ ...node, required: true }),
        defaultTo: x => ({ ...node, default: x }),
    };
    return node.validate(isObject, 'Not an object');
};

export const objectOf = <T extends AnyShapeTomato>(val: T): RecordTomato<T> => {
    const node: RecordTomato<T> = {
        item: val,
        shape: TomatoShape.Record,
        flow: [],
        required: false,
        default: undefined,
        validate: (validate, message = '') => ({
            ...node,
            flow: [...node.flow, { message, validate, type: FlowType.Validate }],
        }),
        require: () => ({ ...node, required: true }),
        defaultTo: x => ({ ...node, default: x }),
    };
    return node.validate(isObject, 'Not an object');
};

export const breeds: Breeds = {
    string,
    number,
    boolean,
    object,
    // TODO:
    objectOf: objectOf as any,
    array,
    any,
};
