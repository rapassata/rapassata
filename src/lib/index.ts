import { Values } from './helpers';
import { Breeds } from './tomatos';

const { any, number, array, boolean, object, objectOf, string } = ('' as any) as Breeds;

// const foo: AtomShapedTomato<string, false> = string;

// type x = UnwrapTomato<typeof string>
const b = object({ o: string, r: string.require() });
type B = Values<typeof b>;

type A = Values<typeof string>;

// const c = object({ foo: array(number) });
const c = array(number);
type C = Values<typeof c>;

const Nest = object({
    age: number.defaultTo(18),
    user: object({
        address: object({
            street: string.validate(x => x !== ''),
            number: string.require(),
        }).require(),
        s: string.defaultTo('red'),
    }),
});
type Y = Values<typeof Nest>;

const Req = {
    data: any,
    user: {
        age: number.defaultTo(18),
        address: {
            street: string,
            number: string.require(),
        },
        colors: array(string.defaultTo('red')).defaultTo(''),
        ratedHobbies: objectOf(number).require(),
    },
};

type X = Values<typeof Req>;
