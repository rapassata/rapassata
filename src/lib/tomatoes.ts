import { TomatoToValues } from "./helpers";
import { FlowItem } from "./validate";

export const enum TomatoShape {
    Atom = 'Atom',
    Array = 'Array',
    Object = 'Object',
    Record = 'Record',
    Or = 'Or',
}

export type BakeTomato<T, R, S> = S extends TomatoShape.Atom ? AtomTomato<T, R> :
    S extends TomatoShape.Array ? ArrayTomato<T, R> :
    S extends TomatoShape.Object ? ObjectTomato<T, R> :
    S extends TomatoShape.Record ? RecordTomato<T, R> :
    S extends TomatoShape.Or ? OrTomato<T, R> :
    never;

export type UnbakeTomato<T> = T extends AtomTomato<infer U, infer R>
    ? BaseTomato<U, R, TomatoShape.Atom>
    : T extends ArrayTomato<infer U, infer R>
    ? BaseTomato<U, R, TomatoShape.Array>
    : T extends ObjectTomato<infer U, infer R>
    ? BaseTomato<U, R, TomatoShape.Object>
    : T extends RecordTomato<infer U, infer R>
    ? BaseTomato<U, R, TomatoShape.Record>
    : T;

type ReturnToShape<Y> = Y extends Array<infer I> ? TomatoShape.Array :
    Y extends object ? TomatoShape.Object : TomatoShape.Atom;

export interface BaseTomato<T, R, S extends TomatoShape> {
    shape: S;
    required: boolean;
    default?: TomatoToValues<T, R, S>;
    flow: Array<FlowItem>;

    require: () => BakeTomato<T, true, S>;
    defaultTo: (defVal: TomatoToValues<T, R, S>) => BakeTomato<T, true, S>;
    // transform: <Y>(t: (x: TomatoToValues<T, R, S>) => Y) => BakeTomato<Y, R, ReturnToShape<Y>>;
    transform: (t: (x: TomatoToValues<T, R, S>) => T) => BakeTomato<T, R, S>;
    validate: (fn: (val: TomatoToValues<T, R, S>) => Promise<boolean> | boolean, message?: string) => BakeTomato<T, R, S>;
}

type s = BakeTomato<boolean | string, true, TomatoShape.Or>

export interface AtomTomato<T, R = false> extends BaseTomato<T, R, TomatoShape.Atom> {
    shape: TomatoShape.Atom;
}
export interface ArrayTomato<T, R = false> extends BaseTomato<T, R, TomatoShape.Array> {
    shape: TomatoShape.Array;
    item: T;
}
export interface ObjectTomato<T, R = false> extends BaseTomato<T, R, TomatoShape.Object> {
    structure: T;
    shape: TomatoShape.Object;
}
export interface RecordTomato<T, R = false> extends BaseTomato<T, R, TomatoShape.Record> {
    item: T;
    shape: TomatoShape.Record;
}
export interface OrTomato<T, R = false> extends BaseTomato<T, R, TomatoShape.Or> {
    shape: TomatoShape.Or;
    fn: 'or';
}

export type Tomato<T, R = false> = AtomTomato<T, R> | ArrayTomato<T, R> | ObjectTomato<T, R> | RecordTomato<T, R> | OrTomato<T, R>;

// Builders
export interface Breeds {
    string: AtomTomato<string>;
    number: AtomTomato<number>;
    boolean: AtomTomato<boolean>;
    any: AtomTomato<any>;
    array: <T extends Tomato<any, any>>(x: T) => ArrayTomato<T>;
    objectOf: <V extends Tomato<any, any>>(val: V) => RecordTomato<Record<any, V>>;
    object: <V>(val: V) => ObjectTomato<V>;
}

const x = '' as any as AtomTomato<string>;
x.transform(x =>)
x.validate(x =>)