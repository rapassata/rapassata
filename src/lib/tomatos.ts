import { Values, Values2 } from './helpers';
import { FlowItem } from './validate';

export const enum TomatoShape {
    Atom = 'Atom',
    Array = 'Array',
    Object = 'Object',
    Record = 'Record',
}

export interface Tomato<
T,
R extends boolean = false,
S extends TomatoShape = TomatoShape.Atom,
BakedTomato = S extends TomatoShape.Atom ? AtomShapedTomato<T, R>
: S extends TomatoShape.Array ? ArrayShapedTomato<T, R>
: S extends TomatoShape.Object ? ObjectShapedTomato<T, R>
: S extends TomatoShape.Record ? RecordShapedTomato<T, R>
: T,
TrueBakedTomato = S extends TomatoShape.Atom ? AtomShapedTomato<T, true>
: S extends TomatoShape.Array ? ArrayShapedTomato<T, true>
: S extends TomatoShape.Object ? ObjectShapedTomato<T, true>
: S extends TomatoShape.Record ? RecordShapedTomato<T, true>
: T,
> {
    shape: TomatoShape;
    required: boolean;
    default?: Values2<T, T, S>;
    flow: Array<FlowItem>;

    // modeling methods
    require: () => TrueBakedTomato;
    defaultTo: (defVal: Values2<T, T, S>) => TrueBakedTomato;
    validate: (fn: (val: Values2<T, T, S>) => boolean | Promise<boolean>, message?: string) => BakedTomato;
}


// Tomato variances
export interface AtomShapedTomato<T, R extends boolean = false> extends Tomato<T, R, TomatoShape.Atom> {
    shape: TomatoShape.Atom;
}

export interface ArrayShapedTomato<T, R extends boolean = false> extends Tomato<T, R, TomatoShape.Array> {
    shape: TomatoShape.Array;
    item: AnyShapeTomato;
}

export interface ObjectShapedTomato<T, R extends boolean = false> extends Tomato<T, R, TomatoShape.Object> {
    shape: TomatoShape.Object;
    structure: T;
}

export interface RecordShapedTomato<T, R extends boolean = false> extends Tomato<T, R, TomatoShape.Record> {
    shape: TomatoShape.Record;
    structure: Record<any, T>;
}

export type AnyShapeTomato<T = any, R extends boolean = any> =
    | AtomShapedTomato<T, R>
    | ArrayShapedTomato<T, R>
    | ObjectShapedTomato<T, R>
    | RecordShapedTomato<T, R>;

type BakeShape<T> = T extends Tomato<infer U, infer R, TomatoShape.Atom> ? AtomShapedTomato<U, R>
: T extends Tomato<infer U, infer R, TomatoShape.Array> ? ArrayShapedTomato<U, R>
: T extends Tomato<infer U, infer R, TomatoShape.Object> ? ObjectShapedTomato<U, R>
: T extends Tomato<infer U, infer R, TomatoShape.Record> ? RecordShapedTomato<U, R>
: T

export type ArrayTomato<T extends AnyShapeTomato = any> = ArrayShapedTomato<T, false>;
export type ObjectTomato<T = any> = ObjectShapedTomato<T, false>;
export type RecordTomato<T extends Record<any, any>> = RecordShapedTomato<T, false>;

// Builders
export interface Breeds {
    string: AtomShapedTomato<string>;
    number: AtomShapedTomato<number>;
    boolean: AtomShapedTomato<boolean>;
    any: AtomShapedTomato<any>;
    array: <T extends AnyShapeTomato>(x: T) => ArrayTomato<T>;
    objectOf: <V>(val: V) => RecordTomato<Record<any, V>>;
    object: <V>(val: V) => ObjectTomato<V>;
}
