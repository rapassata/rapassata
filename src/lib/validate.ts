import { AnyShapeTomato, ArrayTomato, AtomShapedTomato, ObjectTomato, TomatoShape } from './tomatos';

enum FlowType {
    Transform = 'Transform',
    Validate = 'Validate',
}

interface TransformElement<I, O> {
    type: FlowType.Transform;
    transform: (x: I) => O;
}

interface ValidationElement<T> {
    type: FlowType.Validate;
    validate: (x: T) => Promise<Boolean> | boolean;
    message: string;
}

interface ValidationError {
    message: string;
    value: any;
}

export type FlowItem<T = any, O = any> = TransformElement<T, O> | ValidationElement<T>;

const baseNode = <B extends { shape: TomatoShape.Atom }, T>(base: B) => {
    const node: AtomShapedTomato<T> = {
        ...base,
        flow: [],
        required: false,
        default: undefined,
        validate: (validate, message = '') => ({
            ...node,
            flow: [...node.flow, { message, validate, type: FlowType.Validate }],
        }),
        require: () => ({ ...node, required: true }),
        defaultTo: x => ({ ...node, default: x }),
    };
    return node;
};

const atomNode = <T>() => {
    const node: AtomShapedTomato<T> = baseNode({
        shape: TomatoShape.Atom,
    });
    return node;
};

export const string: AtomShapedTomato<string> = (() => {
    const node = atomNode<string>();
    return node.validate(x => typeof x === 'string', 'Not a string');
})();

export const any: AtomShapedTomato<any> = (() => {
    return atomNode<any>();
})();

export const number: AtomShapedTomato<number> = (() => {
    const node = atomNode<number>();
    return node.validate(x => typeof x === 'number', 'Not a number');
})();

export const boolean: AtomShapedTomato<boolean> = (() => {
    const node = atomNode<boolean>();
    return node.validate(x => typeof x === 'boolean', 'Not a boolean');
})();

export const array = <T extends AnyShapeTomato>(item: T): ArrayTomato<T> => {
    const node: ArrayTomato<T> = {
        item,
        shape: TomatoShape.Array,
        flow: [],
        required: false,
        default: undefined,
        validate: (validate, message = '') => ({
            ...node,
            flow: [...node.flow, { message, validate, type: FlowType.Validate }],
        }),
        require: () => ({ ...node, required: true }),
        defaultTo: x => ({ ...node, default: x }),
    };
    return node.validate(x => Array.isArray(x), 'Not an array');
};

const isObject = (x: any) => typeof x === 'object' && x && x.constructor === Object;

export const object = <T>(structure: T): ObjectTomato<T> => {
    const node: ObjectTomato<T> = {
        structure,
        shape: TomatoShape.Object,
        flow: [],
        required: false,
        default: undefined,
        validate: (validate, message = '') => ({
            ...node,
            flow: [...node.flow, { message, validate, type: FlowType.Validate }],
        }),
        require: () => ({ ...node, required: true }),
        defaultTo: x => ({ ...node, default: x }),
    };
    return node.validate(isObject, 'Not an object');
};

const runFlow = (flow: FlowItem[], value: any) => {
    const errors: ValidationError[] = [];
    flow.reduce((value, flowItem) => {
        if (flowItem.type === FlowType.Validate && !flowItem.validate(value)) {
            errors.push({ value, message: flowItem.message });
        }
        return value;
    }, value);
    return {
        errors,
        value,
    };
};

export const validate = (schema: AnyShapeTomato, value: any): any => {
    const flowRes = runFlow(schema.flow, value);
    if (schema.shape === TomatoShape.Atom) {
        return flowRes;
    }
    if (schema.shape === TomatoShape.Array) {
        return {
            ...flowRes,
            children: (Array.isArray(value) ? value : []).map(v => validate(schema.item, v)),
        };
    }
    value = isObject(value) ? value : {};
    return {
        ...flowRes,
        structure: Object.keys(schema.structure).reduce(
            (res, key) => {
                res[key] = validate(schema.structure[key], value[key]);
                return res;
            },
            {} as any
        ),
    };
};
