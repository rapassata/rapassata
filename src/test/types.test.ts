import { inspectWithPreamble } from 'intspector';

describe('Types', () => {
    test('composite', () => {
        const res = inspectWithPreamble(
            `import { Values } from 'lib/helpers';
            import { Breeds } from 'lib/tomatos';
            const { any, number, array, boolean, object, objectOf, string } = ('' as any) as Breeds;
            const Foo = object({
                age: number.defaultTo(18),
                user: object({
                    address: object({
                        street: string.validate(x => x !== ''),
                        number: string.require(),
                    }).require(),
                    s: string.defaultTo('red'),
                }),
                address: object({
                    street: string.validate(x => x !== ''),
                    number: string.require(),
                }).require(),
            });`
        )('Values<typeof Foo>;');
        expect(res).toMatchSnapshot();
    });
});
