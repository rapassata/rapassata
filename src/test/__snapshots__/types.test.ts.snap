// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Types composite 1`] = `"{ age: number; user: { address: { street: string | undefined; number: string; }; s: string; } | undefined; address: { street: string | undefined; number: string; }; } | undefined"`;
